runtime/s390x.S

```diff
@@ -273,68 +614,334 @@ caml_start_program:
+    /* Create an exception handler on the target stack
+       after 16byte DWARF & gc_regs block (which is unused here) */
+        lay     %r9, -32(%r9)
+        larl    TMP, LBL(fiber_exn_handler)
+        stg     TMP, 8(%r9)
+    /* link the previous exn_handler so that copying stacks works */
+        lg      TMP, Stack_exception(%r2)
+        stg     TMP, 0(%r9)
+        lgr     TRAP_PTR, %r9
+    /* Switch to the new stack */
+        lgr     %r15, %r9
+#ifdef ASM_CFI_SUPPORTED
+        CFI_REMEMBER_STATE
+        .cfi_escape DW_CFA_def_cfa_expression, 3+3+2, \
+          DW_OP_breg + DW_REG_r15,                    \
+          caml_runstack_sleb128_1byte,                \
+          DW_OP_deref,                                \
+          DW_OP_plus_uconst, Stack_sp, DW_OP_deref,   \
+          DW_OP_plus_uconst, RETADDR_ENTRY_SIZE
+#endif
+    /* Call the function on the new stack */
+        lgr     %r2, %r4 /* first argument */
+        basr    %r14, %r5 /* closure in %r3 (second argument) */
+LBL(frame_runstack):
+        lay     %r8, 32(%r15) /*%r8 = stack_handler */
+        lg      %r7, Handler_value(%r8) /* saved across C call */
+LBL(caml_runstack_1):
+        lgr     %r12, %r2 /* save return across C call */
+        lg      %r2, Caml_state(current_stack)
+    /* restore parent stack and exn_handler into Caml_state */
+        lg      TMP, Handler_parent(%r8)
+        stg     TMP, Caml_state(current_stack)
+        lg      TRAP_PTR, Stack_exception(TMP)
+        stg     TRAP_PTR, Caml_state(exn_handler)
+    /* free old stack by switching directly to c_stack; is a no-alloc call */
+        lg      %r9, Stack_sp(TMP) /* saved across C call */
+        CFI_RESTORE_STATE
+        CFI_REMEMBER_STATE
+        CFI_DEF_CFA_REGISTER(DW_REG_r9)
+        lg      %r15, Caml_state(c_stack)
+        PREPARE_FOR_C_CALL
+        CFI_ADJUST(160)
+        CHECK_STACK_ALIGNMENT
+        brasl %r14, GCALL(caml_free_stack)
+        CLEANUP_AFTER_C_CALL
+    /* switch directly to parent stack with correct return */
+        lgr     %r2,  %r12
+        lgr     %r3,  %r7
+        lgr     %r15, %r9
+        CFI_RESTORE_STATE
+        lg      TMP, 0(%r3) /* code pointer */
+    /* Invoke handle_value (or handle_exn) */
+        lg      %r14, 0(%r15)
+        CFI_RESTORE(14)
+        la      %r15, 8(%r15)
+        br      TMP
+LBL(fiber_exn_handler):
+        lay     %r8, 16(%r15)
+        lg      %r7, Handler_exception(%r8)
+        brcl    15, LBL(caml_runstack_1)
+CFI_ENDPROC
+ENDFUNCTION(G(caml_runstack))
+
+FUNCTION(G(caml_ml_array_bound_error))
+CFI_STARTPROC
+        lay     %r15, -8(%r15)
+        CFI_ADJUST(8)
+        stg     %r14, 0(%r15)
+        CFI_OFFSET(14, -168)
+        ENTER_FUNCTION
+        LEA_VAR(caml_array_bound_error_asm, ADDITIONAL_ARG)
+        brcl    15, LBL(caml_c_call)
+CFI_ENDPROC
+ENDFUNCTION(G(caml_ml_array_bound_error))
+
+FUNCTION(G(caml_assert_stack_invariants))
```

Comment:
What's the purpose of this `caml_assert_stack_invariants` definition?  It seems unused as nothing emits it, and just calls caml_array_bound_error_asm.

Reply:
Good catch. In the debug mode in the `amd64` backend, a call is emitted to `caml_assert_stack_invariants` at function entry to check stack invariants:

https://github.com/ocaml/ocaml/blob/23dab79a4e42856aa33816b9c79c3d4d79959cb9/asmcomp/amd64/emit.mlp#L901-L902

Here is the implementation of `caml_assert_stack_invariants` on amd64:

https://github.com/ocaml/ocaml/blob/23dab79a4e42856aa33816b9c79c3d4d79959cb9/runtime/amd64.S#L1035-L1047

This functionality is optional; `arm64` and `riscv` don't implement it. I'd recommend that this function be removed.

Reply:
Thanks! Yeah, I'll remove it.